src/portable/renesas/link/dcd_link.c

```diff
@@ -0,0 +1,826 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * This file is part of the TinyUSB stack.
+ */
+
+#include "tusb_option.h"
+
+// Since TinyUSB doesn't use SOF for now, and this interrupt too often (1ms interval)
+// We disable SOF for now until needed later on
+#define USE_SOF 0
+
+#if CFG_TUD_ENABLED && (CFG_TUSB_MCU == OPT_MCU_RX63X || CFG_TUSB_MCU == OPT_MCU_RX65X || CFG_TUSB_MCU == OPT_MCU_RX72N || \
+			CFG_TUSB_MCU == OPT_MCU_RAXXX)
+
+#include "device/dcd.h"
+#include "link_type.h"
+
+#if TU_CHECK_MCU(OPT_MCU_RX63X, OPT_MCU_RX65X, OPT_MCU_RX72N)
+#include "link_rx.h"
+#elif TU_CHECK_MCU(OPT_MCU_RAXXX)
+#include "link_ra.h"
+#else
+#error "Unsupported MCU"
+#endif
+
+//--------------------------------------------------------------------+
+// MACRO TYPEDEF CONSTANT ENUM
+//--------------------------------------------------------------------+
+
+/* LINK core registers */
+#define LINK_REG ((LINK_REG_t *) LINK_REG_BASE)
+
+/* Start of definition of packed structs (used by the CCRX toolchain) */
+TU_ATTR_PACKED_BEGIN
+TU_ATTR_BIT_FIELD_ORDER_BEGIN
+
+typedef struct TU_ATTR_PACKED {
+	union {
+		struct {
+			uint16_t : 8;
+			uint16_t TRCLR : 1;
+			uint16_t TRENB : 1;
+			uint16_t : 0;
+		};
+		uint16_t TRE;
+	};
+	uint16_t TRN;
+} reg_pipetre_t;
+
+typedef union TU_ATTR_PACKED {
+	struct {
+		volatile uint16_t u8 : 8;
+		volatile uint16_t : 0;
+	};
+	volatile uint16_t u16;
+} hw_fifo_t;
+
+typedef struct TU_ATTR_PACKED {
+	void *buf;	    /* the start address of a transfer data buffer */
+	uint16_t length;    /* the number of bytes in the buffer */
+	uint16_t remaining; /* the number of bytes remaining in the buffer */
+	struct {
+		uint32_t ep : 8; /* an assigned endpoint address */
+		uint32_t ff : 1; /* `buf` is TU_FUFO or POD */
+		uint32_t : 0;
+	};
+} pipe_state_t;
+
+TU_ATTR_PACKED_END /* End of definition of packed structs (used by the CCRX toolchain) */
+	TU_ATTR_BIT_FIELD_ORDER_END
+
+	typedef struct {
```

Comment:
This indentations may be typos.

Reply:
yes, the formatter I'm using (clang-format) doesnt like those